### PR TITLE
Support getBlobChunkIds API for composite blob

### PR DESCRIPTION
**Pull request by SophieGuo410**
Will set it as admin request in close source.

### CODE REVIEW

**Review by snalli — COMMENTED** [review]
Can you add some context in the commit msg as to why this is needed ? Paste the GCN number if u have one

File: ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouter.java [hunk 1]
```diff
@@ -69,6 +70,7 @@ class NonBlockingRouter implements Router {
   private final Cache<String, Boolean> notFoundCache;
   private final AmbryCache blobMetadataCache;
 
+
```
Comment (snalli) [inline ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouter.java:73]: rm newline

File: ambry-router/src/test/java/com/github/ambry/router/GetManagerTest.java [hunk 2]
```diff
@@ -534,6 +545,18 @@ private void compareBlobInfo(BlobInfo blobInfo, long segmentSize) {
...
 
+  /**
+   * Compare the size of chunk BlobId list with the expected chunk size + metadata chunk.
+   * @param chunkBlobIds List of chunk blob Ids for the composite blob.
+   */
+  private void compareChunkBlobIds(long blobSize, List<StoreKey> chunkBlobIds) {
+    if (blobSize <= CHUNK_SIZE) {
+      Assert.assertNull("chunkBlobIds should be null for simple blob", chunkBlobIds);
+    } else {
+      Assert.assertEquals("the size of chunk BlobId list should match", blobSize / CHUNK_SIZE + 1, chunkBlobIds.size());
```
Comment (snalli) [inline ambry-router/src/test/java/com/github/ambry/router/GetManagerTest.java:556]: Is this always true ? There could be a 4MB composite blob comprising four 1-MB chunks.

File: ambry-router/src/test/java/com/github/ambry/router/GetBlobOperationTest.java [hunk 3]
```diff
@@ -1743,6 +1743,7 @@ private GetBlobOperation getAndAssertSuccess(final boolean getChunksBeforeRead,
                   blobInfo.getBlobProperties().getBlobSize());
               Assert.assertNull("Unexpected blob data in operation result", result.getBlobResult.getBlobDataChannel());
               Assert.assertEquals("LifeVersion mismatch", expectedLifeVersion, blobInfo.getLifeVersion());
+              break;
```
Comment (snalli) [inline ambry-router/src/test/java/com/github/ambry/router/GetBlobOperationTest.java:1746]: not needed.

File: ambry-frontend/src/main/java/com/github/ambry/frontend/GetBlobHandler.java [hunk 4]
```diff
@@ -162,7 +163,7 @@ private Callback<Void> securityPostProcessRequestCallback(BlobId blobId) {
           RestRequestMetrics restRequestMetrics = metricsGroup.getRestRequestMetrics(restRequest.isSslUsed(), true);
           restRequest.getMetricsTracker().injectMetrics(restRequestMetrics);
         }
-        if (subResource == null) {
+        if (subResource == null || subResource == SubResource.BlobChunkIds) {
```
Comment (justinlin-linkedin) [inline ambry-frontend/src/main/java/com/github/ambry/frontend/GetBlobHandler.java:166]: This is a bit odd, we should probably do something like
```
if (subResource == SubResource.Replicas) {
    finalCallback.onCompletion(getReplicasHandler.getReplicas(blobId.getID(), restResponseChannel), null);
} else {
    router.getBlob(blobId.getID(), options, routerCallback(), QuotaUtils.buildQuotaChargeCallback(restRequest, quotaManager, true));
}
```

File: ambry-api/src/main/java/com/github/ambry/router/GetBlobResult.java [hunk 5]
```diff
@@ -23,15 +25,18 @@
 public class GetBlobResult {
   private final BlobInfo blobInfo;
   private final ReadableStreamChannel blobDataChannel;
+  private final List<StoreKey> blobChunkIds;
```
Comment (justinlin-linkedin) [inline ambry-api/src/main/java/com/github/ambry/router/GetBlobResult.java:28]: we already have a class called GetBlobResultInternal, which includes List of store keys. Though that serves as an internal structure, we should probably remove that class all together now that we have this one.

File: ambry-router/src/main/java/com/github/ambry/router/GetBlobInfoOperation.java [hunk 6]
```diff
@@ -413,7 +413,7 @@ private void handleBody(InputStream payload, MessageMetadata messageMetadata, Me
     if (encryptionKey == null) {
       // if blob is not encrypted, move the state to Complete
       BlobInfo blobInfo = new BlobInfo(serverBlobProperties, userMetadata.array(), messageInfo.getLifeVersion());
-      operationResult = new GetBlobResultInternal(new GetBlobResult(blobInfo, null), null);
+      operationResult = new GetBlobResultInternal(new GetBlobResult(blobInfo, null, null), null);
```
Comment (justinlin-linkedin) [inline ambry-router/src/main/java/com/github/ambry/router/GetBlobInfoOperation.java:416]: This is what I am talking about, should just get rid of GetBlobResultInternal.

File: ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java [hunk 7]
```diff
@@ -312,7 +312,9 @@ private void onChunkOperationComplete(GetChunk chunk) {
           // GetManager remove this operation.
           setOperationCompleted();
           List<StoreKey> chunkIds = e == null && compositeBlobInfo != null ? compositeBlobInfo.getKeys() : null;
-          operationResult = new GetBlobResultInternal(null, chunkIds);
+          operationResult = new GetBlobResultInternal(
+              options.getBlobOptions.getOperationType() == GetBlobOptions.OperationType.BlobChunkIds
```
Comment (justinlin-linkedin) [inline ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:316]: In `GetBlobOptionsInternal` there is a member called `getChunkIdsOnly`, we should probably get rid of that as well now that we have BlobChunkId as SubResource.

**Review by snalli — CHANGES_REQUESTED** [review]
requesting changes

File: ambry-router/src/test/java/com/github/ambry/router/GetManagerTest.java (hunk 2 above)
Comment (SophieGuo410) [reply ambry-router/src/test/java/com/github/ambry/router/GetManagerTest.java:556]: For all these tests used this getBlobAndCompareContent method, both stitched blob and composite blob has larger size than the one chunk. So I use this way to make the test simple. Let me know if you have any other preference.

File: ambry-api/src/main/java/com/github/ambry/router/GetBlobResult.java (hunk 5 above)
Comment (SophieGuo410) [reply ambry-api/src/main/java/com/github/ambry/router/GetBlobResult.java:28]: You are right, I removed the GetBlobResultInternal.

File: ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java (hunk 7 above)
Comment (SophieGuo410) [reply ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java]: I checked the undelete and delete also need this getChunkIdsOnly in GetBlobOptionsInternal, so I just keep using it. Let me know if I miss anything. thx!

**Pull request closed without merging**